File: docs/manual/scala/guide/cluster/code/docs/home/scaladsl/persistence/CassandraReadSideQuery.scala
/*
 * Copyright (C) 2016-2019 Lightbend Inc. <https://www.lightbend.com>
 */

package docs.home.scaladsl.persistence

//#imports
import scala.concurrent.Future
import akka.NotUsed
import akka.stream.scaladsl.Source
import com.lightbend.lagom.scaladsl.api.Service
import com.lightbend.lagom.scaladsl.api.ServiceCall
import com.lightbend.lagom.scaladsl.persistence.cassandra.CassandraSession

//#imports

trait CassandraReadSideQuery {

  trait BlogService extends Service {
    def getPostSummaries(): ServiceCall[NotUsed, Source[PostSummary, _]]

    override def descriptor = ???
  }

  //#service-impl
  class BlogServiceImpl(cassandraSession: CassandraSession) extends BlogService {

    override def getPostSummaries() = ServiceCall { request =>
      val response: Source[PostSummary, NotUsed] =
        cassandraSession
          .select("SELECT id, title FROM blogsummary")
          .map(row => PostSummary(row.getString("id"), row.getString("title")))
      Future.successful(response)
    }
  }
  //#service-impl

}

File: docs/manual/java/guide/cluster/code/docs/home/persistence/CassandraReadSideQuery.java
/*
 * Copyright (C) 2016-2019 Lightbend Inc. <https://www.lightbend.com>
 */

package docs.home.persistence;

// #imports
import akka.NotUsed;
import com.lightbend.lagom.javadsl.api.ServiceCall;
import com.lightbend.lagom.javadsl.persistence.cassandra.CassandraSession;
import java.util.concurrent.CompletableFuture;
import javax.inject.Inject;
import akka.stream.javadsl.Source;
// #imports

public interface CassandraReadSideQuery {

  interface BlogService {
    ServiceCall<NotUsed, Source<PostSummary, ?>> getPostSummaries();
  }

  // #service-impl
  public class BlogServiceImpl implements BlogService {

    private final CassandraSession cassandraSession;

    @Inject
    public BlogServiceImpl(CassandraSession cassandraSession) {
      this.cassandraSession = cassandraSession;
    }

    @Override
    public ServiceCall<NotUsed, Source<PostSummary, ?>> getPostSummaries() {
      return request -> {
        Source<PostSummary, ?> summaries =
            cassandraSession
                .select("SELECT id, title FROM blogsummary;")
                .map(row -> new PostSummary(row.getString("id"), row.getString("title")));
        return CompletableFuture.completedFuture(summaries);
      };
    }
  }
  // #service-impl

}
